ccl/changefeedccl: TestChangefeedColumnFamily failed

ccl/changefeedccl.TestChangefeedColumnFamily [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3613638&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3613638&tab=artifacts#/) on master @ [89948a5f685b0fe274df4720745401e3f3f64093](https://github.com/cockroachdb/cockroach/commits/89948a5f685b0fe274df4720745401e3f3f64093):


```
=== RUN   TestChangefeedColumnFamily
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestChangefeedColumnFamily044463890
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestChangefeedColumnFamily
    changefeed_test.go:1350: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestChangefeedColumnFamily044463890
--- FAIL: TestChangefeedColumnFamily (7.28s)
=== RUN   TestChangefeedColumnFamily/sinkless
    changefeed_test.go:1342: expected "exactly 1 column family" error got: <nil>
    --- FAIL: TestChangefeedColumnFamily/sinkless (1.25s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
danhhz marked as alumn{us,a}; ignoring
```
</p>
</details>

/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedColumnFamily.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -444,17 +444,22 @@ func (ca *changeAggregator) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMet
 			return ca.ProcessRowHelper(ca.resolvedSpanBuf.Pop()), nil
 		}
 		if err := ca.tick(); err != nil {
-			if errors.Is(err, kvevent.ErrBufferClosed) {
+			var e kvevent.ErrBufferClosed
+			if errors.As(err, &e) {
 				// ErrBufferClosed is a signal that
 				// our kvfeed has exited expectedly.
-				err = nil
-			}
+				err = e.Unwrap()
+				if errors.Is(err, kvevent.ErrNormalRestartReason) {
+					err = nil
+				}
+			} else {
 
-			select {
-			// If the poller errored first, that's the
-			// interesting one, so overwrite `err`.
-			case err = <-ca.errCh:
-			default:
+				select {
+				// If the poller errored first, that's the
+				// interesting one, so overwrite `err`.
+				case err = <-ca.errCh:
+				default:
+				}
 			}
 			// Shut down the poller if it wasn't already.
 			ca.cancel()

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -1315,7 +1315,6 @@ func TestChangefeedAfterSchemaChangeBackfill(t *testing.T) {
 
 func TestChangefeedColumnFamily(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 71796, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	testFn := func(t *testing.T, db *gosql.DB, f cdctest.TestFeedFactory) {
@@ -1464,7 +1463,6 @@ func requireErrorSoon(
 func TestChangefeedFailOnTableOffline(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.UnderDeadlock(t, "timeout")
 
 	dataSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		if r.Method == "GET" {

--- a/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go
+++ b/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go
@@ -34,6 +34,7 @@ type blockingBuffer struct {
 	mu struct {
 		syncutil.Mutex
 		closed  bool             // True when buffer closed.
+		reason  error            // Reason buffer is closed.
 		drainCh chan struct{}    // Set when Drain request issued.
 		blocked bool             // Set when event is blocked, waiting to acquire quota.
 		queue   bufferEntryQueue // Queue of added events.
@@ -74,7 +75,7 @@ func (b *blockingBuffer) pop() (e *bufferEntry, err error) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 	if b.mu.closed {
-		return nil, ErrBufferClosed
+		return nil, ErrBufferClosed{reason: b.mu.reason}
 	}
 
 	e = b.mu.queue.dequeue()
@@ -240,8 +241,8 @@ func (b *blockingBuffer) Drain(ctx context.Context) error {
 	return nil
 }
 
-// Close implements Writer interface.
-func (b *blockingBuffer) Close(ctx context.Context) error {
+// CloseWithReason implements Writer interface.
+func (b *blockingBuffer) CloseWithReason(ctx context.Context, reason error) error {
 	// Close quota pool -- any requests waiting to acquire will receive an error.
 	b.qp.Close("blocking buffer closing")
 
@@ -276,6 +277,7 @@ func (b *blockingBuffer) Close(ctx context.Context) error {
 	}
 
 	b.mu.closed = true
+	b.mu.reason = reason
 	close(b.signalCh)
 
 	// Return all queued up entries to the buffer pool.

--- a/pkg/ccl/changefeedccl/kvevent/blocking_buffer_test.go
+++ b/pkg/ccl/changefeedccl/kvevent/blocking_buffer_test.go
@@ -80,7 +80,7 @@ func TestBlockingBuffer(t *testing.T) {
 	st := cluster.MakeTestingClusterSettings()
 	buf := kvevent.NewMemBuffer(ba, &st.SV, &metrics, quotapool.OnWaitStart(notifyWait))
 	defer func() {
-		require.NoError(t, buf.Close(context.Background()))
+		require.NoError(t, buf.CloseWithReason(context.Background(), nil))
 	}()
 
 	producerCtx, stopProducers := context.WithCancel(context.Background())
@@ -123,7 +123,7 @@ func TestBlockingBufferNotifiesConsumerWhenOutOfMemory(t *testing.T) {
 	st := cluster.MakeTestingClusterSettings()
 	buf := kvevent.NewMemBuffer(ba, &st.SV, &metrics)
 	defer func() {
-		require.NoError(t, buf.Close(context.Background()))
+		require.NoError(t, buf.CloseWithReason(context.Background(), nil))
 	}()
 
 	producerCtx, stopProducer := context.WithCancel(context.Background())

--- a/pkg/ccl/changefeedccl/kvevent/chan_buffer.go
+++ b/pkg/ccl/changefeedccl/kvevent/chan_buffer.go
@@ -17,7 +17,8 @@ import (
 // chanBuffer mediates between the changed data KVFeed and the rest of the
 // changefeed pipeline (which is backpressured all the way to the sink).
 type chanBuffer struct {
-	entriesCh chan Event
+	entriesCh    chan Event
+	closedReason error
 }
 
 // MakeChanBuffer returns an Buffer backed by an unbuffered channel.
@@ -45,7 +46,9 @@ func (b *chanBuffer) Drain(ctx context.Context) error {
 	return nil
 }
 
-func (b *chanBuffer) Close(_ context.Context) error {
+// CloseWithReason implements Writer interface.
+func (b *chanBuffer) CloseWithReason(_ context.Context, reason error) error {
+	b.closedReason = reason
 	close(b.entriesCh)
 	return nil
 }
@@ -60,7 +63,7 @@ func (b *chanBuffer) Get(ctx context.Context) (Event, error) {
 		if !ok {
 			// Our channel has been closed by the
 			// Writer. No more events will be returned.
-			return e, ErrBufferClosed
+			return e, ErrBufferClosed{reason: b.closedReason}
 		}
 		e.bufferGetTimestamp = timeutil.Now()
 		return e, nil

--- a/pkg/ccl/changefeedccl/kvevent/event.go
+++ b/pkg/ccl/changefeedccl/kvevent/event.go
@@ -23,7 +23,22 @@ import (
 
 // ErrBufferClosed is returned by Readers when no more values will be
 // returned from the buffer.
-var ErrBufferClosed = errors.New("buffer closed")
+type ErrBufferClosed struct {
+	reason error
+}
+
+// Error() implements the error interface
+func (e ErrBufferClosed) Error() string {
+	return "buffer closed"
+}
+
+func (e ErrBufferClosed) Unwrap() error {
+	return e.reason
+}
+
+// ErrNormalRestartReason is a sentinel error to indicate the
+// ErrBufferClosed's reason is a pending restart
+var ErrNormalRestartReason = errors.New("writer can restart")
 
 // Buffer is an interface for communicating kvfeed entries between processors.
 type Buffer interface {
@@ -43,8 +58,8 @@ type Writer interface {
 	Add(ctx context.Context, event Event) error
 	// Drain waits until all events buffered by this writer has been consumed.
 	Drain(ctx context.Context) error
-	// Close closes this writer.
-	Close(ctx context.Context) error
+	// CloseWithReason closes this writer. reason may be added as a detail to ErrBufferClosed.
+	CloseWithReason(ctx context.Context, reason error) error
 }
 
 // Type indicates the type of the event.

--- a/pkg/ccl/changefeedccl/kvfeed/kv_feed.go
+++ b/pkg/ccl/changefeedccl/kvfeed/kv_feed.go
@@ -114,7 +114,7 @@ func Run(ctx context.Context, cfg Config) error {
 	if !errors.As(err, &scErr) {
 		// Regardless of whether we exited KV feed with or without an error, that error
 		// is not a schema change; so, close the writer and return.
-		return errors.CombineErrors(err, f.writer.Close(ctx))
+		return errors.CombineErrors(err, f.writer.CloseWithReason(ctx, err))
 	}
 
 	log.Infof(ctx, "stopping kv feed due to schema change at %v", scErr.ts)
@@ -124,7 +124,7 @@ func Run(ctx context.Context, cfg Config) error {
 	// Regardless of whether drain succeeds, we must also close the buffer to release
 	// any resources, and to let the consumer (changeAggregator) know that no more writes
 	// are expected so that it can transition to a draining state.
-	err = errors.CombineErrors(f.writer.Drain(ctx), f.writer.Close(ctx))
+	err = errors.CombineErrors(f.writer.Drain(ctx), f.writer.CloseWithReason(ctx, kvevent.ErrNormalRestartReason))
 
 	if err == nil {
 		// This context is canceled by the change aggregator when it receives
@@ -352,7 +352,7 @@ func (f *kvFeed) runUntilTableEvent(
 
 	memBuf := f.bufferFactory()
 	defer func() {
-		err = errors.CombineErrors(err, memBuf.Close(ctx))
+		err = errors.CombineErrors(err, memBuf.CloseWithReason(ctx, err))
 	}()
 
 	g := ctxgroup.WithContext(ctx)

--- a/pkg/ccl/changefeedccl/kvfeed/scanner_test.go
+++ b/pkg/ccl/changefeedccl/kvfeed/scanner_test.go
@@ -42,7 +42,7 @@ func (r *recordResolvedWriter) Drain(ctx context.Context) error {
 	return nil
 }
 
-func (r *recordResolvedWriter) Close(ctx context.Context) error {
+func (r *recordResolvedWriter) CloseWithReason(ctx context.Context, reason error) error {
 	return nil
 }
 

